python/cudf/bindings/reduce.pyx

```diff
@@ -15,57 +15,107 @@ import pandas as pd
 
-
 from librmm_cffi import librmm as rmm
 
-
 from libc.stdint cimport uintptr_t
 from libc.stdlib cimport calloc, malloc, free
 
 from libcpp.map cimport map as cmap
 from libcpp.string  cimport string as cstring
 
 
-
-def apply_reduce(reduction, col):
+_REDUCTION_OP = {
+  'max': GDF_REDUCTION_MAX,
+  'min': GDF_REDUCTION_MIN,
+  'sum': GDF_REDUCTION_SUM,
+  'product': GDF_REDUCTION_PRODUCT,
+  'sum_of_squares': GDF_REDUCTION_SUMOFSQUARES,
+}
+
+_SCAN_OP = {
+  'sum': GDF_SCAN_SUM,
+  'min': GDF_SCAN_MIN,
+  'max': GDF_SCAN_MAX,
+  'product': GDF_SCAN_PRODUCT,
+}
+
+
+cdef get_scalar_value(gdf_scalar scalar):
+    return {
+        GDF_FLOAT64: scalar.data.fp64,
+        GDF_FLOAT32: scalar.data.fp32,
+        GDF_INT64:   scalar.data.si64,
+        GDF_INT32:   scalar.data.si32,
+        GDF_INT16:   scalar.data.si16,
+        GDF_INT8:    scalar.data.si08,
+        GDF_DATE32:  scalar.data.dt32,
+        GDF_DATE64:  scalar.data.dt64,
+        GDF_TIMESTAMP: scalar.data.tmst,
+    }[scalar.dtype]
+
+
+def apply_reduce(reduction_op, col, dtype=None):
     """
       Call gdf reductions.
     """
 
+    check_gdf_compatibility(col)
+
+    # check empty case
+    if col.data.size <= col.null_count :
+        if reduction_op == 'sum' or reduction_op == 'sum_of_squares':
+            return col.dtype.type(0)
+        if reduction_op == 'product' and pandas_version >= (0, 22):
+            return col.dtype.type(1)
+        return np.nan
 
-    outsz = gdf_reduction_get_intermediate_output_size()
-    out = rmm.device_array(outsz, dtype=col.dtype)
-    cdef uintptr_t out_ptr = get_ctype_ptr(out)
+    col_dtype = dtype if dtype != None else col.dtype
 
-    check_gdf_compatibility(col)
     cdef gdf_column* c_col = column_view_from_column(col)
+    cdef gdf_reduction_op c_op = _REDUCTION_OP[reduction_op]
+    cdef gdf_dtype c_out_dtype = get_dtype(col_dtype.type if dtype is None else col_dtype)
+    cdef gdf_scalar c_result
+
+    print('c_out_dtype:', c_out_dtype)
 
-    cdef gdf_error result
     with nogil:    
-        if reduction == 'max':
-            result = gdf_max(<gdf_column*>c_col, <void*>out_ptr, outsz)
-        elif reduction == 'min':
-            result = gdf_min(<gdf_column*>c_col, <void*>out_ptr, outsz)
-        elif reduction == 'sum':
-            result = gdf_sum(<gdf_column*>c_col, <void*>out_ptr, outsz)
-        elif reduction == 'sum_of_squares':
-            result = gdf_sum_of_squares(<gdf_column*>c_col,
-                                        <void*>out_ptr,
-                                        outsz)
-        elif reduction == 'product':
-            result = gdf_product(<gdf_column*>c_col, <void*>out_ptr, outsz)
-        else:
-            result = GDF_NOTIMPLEMENTED_ERROR
+        c_result = reduction(
+            <gdf_column*>c_col,
+            c_op,
+            c_out_dtype
+            )
 
     free(c_col)
+    result = np.array([get_scalar_value(c_result)], dtype=col_dtype)
```

Comment:
Good pointing!

The reason I used numpy array was that .astype('f8') was used at NumericalColumn.mean().
  https://github.com/rapidsai/cudf/blob/b63e971f9c2348db6c12882c882139b2e0988d1e/python/cudf/dataframe/numerical.py#L274-L275

I realized I can use numpy.float64(x) instead of .astype('f8')
Returning scalar value no longer needs [0] suffix (vmin = self.min()[0] issue thomcom pointed)
Could you evaluate commit 5685064 and eac4893 ?